Implement `.by` (#6528)
tidyverse/dplyr: man/rmd/by.Rmd

---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

There are two ways to group in dplyr:

-   Persistent grouping with [group_by()]

-   Temporary grouping with `.by`

This help page is dedicated to explaining where and why you might want to use the latter.
The most important reason to use `.by` is that you *never* have to remember to use [ungroup()] after it.
The grouping is always temporary, meaning that if an ungrouped data frame goes in, then an ungrouped data frame comes out, regardless of the number of grouping columns that are specified.

> 

Reviewer: Maybe here would be a good place to summarise the differences between `group_by` and `.by`?

* Grouping only affects a single verb
* Select variables with tidyselect
* Summaries are ordered by appearance

Or maybe a table with one column for `group_by` and one column for `.by`?

Reviewer: It's relatively unimportant but it would be nice to mention somewhere that `df |> summarise(.by)` preserves the class of `df`, and `df |> mutate(.by)` preserves the attributes.

Author: > df |> summarise(.by) preserves the class of df

I think this is only true for classes that `summarise()` knows about. i.e. `summarise()` a data.table and you get a data.frame back

I think I'd prefer to leave these details out for the moment if that is okay with you

Reviewer: Yeah, sure.

The following dplyr verbs support `.by`:

-   [mutate()]

-   [summarise()]

-   [filter()]

-   [slice()] and its variants, such as [slice_head()]

Let's take a look at the two grouping approaches using this `expenses` data set, which tracks costs accumulated across various `id`s and `region`s:

```{r}
expenses <- tibble(
  id = c(1, 2, 1, 3, 1, 2, 3),
  region = c("A", "A", "A", "B", "B", "A", "A"),
  cost = c(25, 20, 19, 12, 9, 6, 6)
)
expenses
```

Imagine that you wanted to compute the average cost per region.
You'd probably write something like this:

```{r}
expenses %>%
  group_by(region) %>%
  summarise(cost = mean(cost))
```

> As of dplyr 1.1.0, an additional option is available that lets you specify the grouping *inline* within the verb:

Reviewer: ```suggestion
Instead, you can now specify the grouping *inline* within the verb:
```

```{r}
expenses %>%
  summarise(cost = mean(cost), .by = region)
```

> This great idea comes from [data.table](https://CRAN.R-project.org/package=data.table), where the equivalent syntax looks something like `expenses[, .(cost = mean(cost)), by = region]`.

Reviewer: Move this up to intro.

Grouping with `.by` is temporary, meaning that since `expenses` was an ungrouped data frame, the result after applying `.by` will also always be an ungrouped data frame, full stop.
Compare that with `group_by() %>% summarise()`, where `summarise()` generally peels off 1 layer of grouping by default, with a message that it is doing so if there were originally more than one grouping columns:

```{r}
expenses %>%
  group_by(id, region) %>%
  summarise(cost = mean(cost))
```

This behavior is sometimes useful to sequentially "roll up" a data frame that has been grouped by multiple columns, but often you just want to group temporarily.
Traditionally, you'd either `ungroup()` after the `summarise()` or explicitly set `.groups = "drop"` to achieve this:

```{r}
expenses %>%
  group_by(id, region) %>%
  summarise(cost = mean(cost), .groups = "drop")
```

Because `.by` grouping is temporary, you don't need to worry about ungrouping and it never needs to emit a message to remind you what is happening:

```{r}
expenses %>%
  summarise(cost = mean(cost), .by = c(id, region))

> ```

Reviewer: I think you could reduce these two paragraphs + code chunks this to a single sentence.

Note that we specified multiple columns to group by using the [tidy-select][dplyr_tidy_select] syntax `c(id, region)`.
If you have a character vector of column names you'd like to group by, you can do so with `.by = all_of(my_cols)`.

To prevent surprising results, you can't use `.by` on an existing grouped data frame:

```{r, error=TRUE}
expenses %>% 
  group_by(id) %>%
  summarise(cost = mean(cost), .by = c(id, region))
```

So far we've focused on the usage of `.by` with `summarise()`, but `.by` works with a number of other dplyr verbs.
For example, you could append the mean cost per region onto the original data frame as a new column rather than computing a summary:

```{r}
expenses %>%
  mutate(cost_by_region = mean(cost), .by = region)
```

Or you could slice out the maximum cost per combination of id and region:

```{r}
expenses %>%
  slice_max(cost, n = 1, by = c(id, region))
```

> Again, note that the result of this `slice_max()` is an ungrouped data frame, which is probably what you'd expect here.

Reviewer: Again, feels a bit too duplicative

Compare that to the `group_by()` approach, which uses persistent grouping and returns another grouped data frame:

```{r}
expenses %>%
  group_by(id, region) %>%
  slice_max(cost, n = 1)
```

### Result ordering

> When used with `.by`, `summarise()` and `slice()` both return results where the group keys are ordered by their existing ordering in the original data rather than by their values.

Reviewer: ```suggestion
When used with `.by`, `summarise()` and `slice()` both maintain ordering of the existing results.
```

This is different from `group_by()`, which has always sorted the keys in ascending order.

```{r}
df <- tibble(
  month = c("jan", "jan", "feb", "feb", "mar"),
  temp = c(20, 25, 18, 20, 40)
)

# Uses ordering by "first appearance" in the original data
df %>%
  summarise(average_temp = mean(temp), .by = month)

# Sorts in ascending order
df %>%
  group_by(month) %>%
  summarise(average_temp = mean(temp))

> # Same with `slice()`

Reviewer: I think you can skip the slice example
